dart/colfer_gen_test.dart

```diff
@@ -0,0 +1,160 @@
+    Golden('0755ef312a2e5da1007f', O(t: DateTime.fromMicrosecondsSinceEpoch(1441739050777888))),
+    Golden('87000007d954159c00000003e87f',
+        O(t: DateTime.fromMicrosecondsSinceEpoch(8630000000000000001))),
+    Golden('87fffff82457de8000000003e87f',
+        O(t: DateTime.fromMicrosecondsSinceEpoch(-8639999999999999999))),
+    Golden('87ffffffffffffffff2e5da1007f', O(t: DateTime.fromMicrosecondsSinceEpoch(-222112))),
+    Golden('0801417f', O(s_: 'A')),
+    Golden('080261007f', O(s_: 'a\x00')),
+    Golden('0809c280e0a080f09080807f', O(s_: '\u0080\u0800\u{10000}')),
+    Golden(
+        '08800120202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020202020207f',
+        O(s_: ' ' * 128)),
+    Golden('0901ff7f', O(a_: Uint8List.fromList([MaxUint8]))),
+    Golden('090202007f', O(a_: Uint8List.fromList([2, 0]))),
+    Golden(
+        '09c0010909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909090909097f',
+        O(a_: tmp)),
+    Golden('0a7f7f', O(o: O())),
+    Golden('0a007f7f', O(o: O(b: true))),
+    Golden('0b01007f7f', O(os: [O(b: true)])),
+    Golden('0b027f7f7f', O(os: [O(), O()])),
+    Golden('0c0300016101627f', O(ss: ['', 'a', 'b'])),
+    Golden(
+        '0d0201000201027f',
+        O(as_: [
+          Uint8List.fromList([0]),
+          Uint8List.fromList([1, 2])
+        ])),
+    Golden('0e017f', O(u8: 1)),
+    Golden('0eff7f', O(u8: MaxUint8)),
+    Golden('8f017f', O(u16: 1)),
+    Golden('0fffff7f', O(u16: MaxUint16)),
+    Golden('1002000000003f8000007f', O(f32s: Float32List.fromList([0, 1]))),
+    Golden('11014058c000000000007f', O(f64s: Float64List.fromList([99]))),
+  ];
+}
+
+void main() {
+  var cases = newGoldenCases();
+  group('Marshal', () {
+    for (final gold in cases) {
+      var buf = BytesBuilder(copy: false);
+      gold.object.marshal(buf);
+      test('marsaled binary should match golden string', () {
+        List<int> list = hex.decode(gold.serial);
+        Uint8List bytes = Uint8List.fromList(list);
+        expect(buf.toBytes(), bytes);
+      });
+    }
+  });
+
+  group('Unmarshal', () {
+    for (final gold in cases) {
+      test('struct should match golden string', () {
+        List<int> list = hex.decode(gold.serial);
+        Uint8List bytes = Uint8List.fromList(list);
+        var obj = O();
+        obj.unmarshal(bytes);
+        // work around NaN != NaN
+        if (obj.f32.isNaN && gold.object.f32.isNaN) {
+          obj.f32 = 0;
+          gold.object.f32 = 0;
```

Comment:
Nice catch!

Reply:
Do you know how to fix this? I really don't like NaN ≠ NaN. No idea why that was decided in the first place either. Maybe we can make the equals do NaN = NaN for all floating points? Do you have an opinion there?

Reply:
I checked, and it's true that NaN ≠ NaN. I can't think of a simpler workaround, and this is totally safe, because the 0 => NaN change occurs if and only if both of them are NaN.

Reply:
We override the equals. In there we could make it return true when both floating points are NaN. In that case you can just assert the object without a workaround, right?

Reply:
> I'll add `RangeError` for `ColferSizeMax` and `ColferListMax`. It's unlikely to happen while unmarshalling, because the same thing should have been run on the other side, so it won't interfere with reading incomplete buffer indeed.

Unlikely and shouldn't yes. Can't rely on that though. The format guarantees robustness.

> "Marshalling and unmarshalling comes with built-in size protection to ensure predictable memory consumption. The format prevents memory bombs by design."
> — README.md

In other words, even if someone tries to manually craft something beyond ColferSizeMax, we should catch that and *not* allocate the size provides by some serial.